include/picongpu/simulation/stage/AtomicPhysics.hpp

```diff
@@ -0,0 +1,604 @@
+            {
+                picongpu::particles::atomicPhysics::stage::DumpSuperCellDataToConsole<
+                    picongpu::particles::atomicPhysics::localHelperFields::LocalTimeStepField<picongpu::MappingDesc>,
+                    picongpu::particles::atomicPhysics::localHelperFields::PrintTimeStepToConsole>{}(
+                    mappingDesc,
+                    "LocalTimeStepField");
+            }
+
+        public:
+            AtomicPhysics() = default;
+
+            //! atomic physics stage sub-stage calls
+            void operator()(picongpu::MappingDesc const mappingDesc, uint32_t const currentStep) const
+            {
+                // debug only
+                using ForEachIonSpeciesDumpRateCacheToConsole = pmacc::meta::ForEach<
+                    AtomicPhysicsIonSpecies,
+                    particles::atomicPhysics::stage::DumpRateCacheToConsole<boost::mpl::_1>>;
+
+                pmacc::DataConnector& dc = pmacc::Environment<>::get().DataConnector();
+
+                // TimeRemainingSuperCellField
+                auto& localTimeRemainingField = *dc.get<S_TimeRemainingField>("LocalTimeRemainingField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutTimeRemaining
+                    = localTimeRemainingField.getGridLayout().sizeWithoutGuardND();
+
+                // FoundUnboundIonSuperCellField
+                auto& localFoundUnboundIonField = *dc.get<S_FoundUnboundField>("LocalFoundUnboundIonField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutFoundUnbound
+                    = localFoundUnboundIonField.getGridLayout().sizeWithoutGuardND();
+
+                // ElectronHistogramOverSubscribedSuperCellField
+                auto& localElectronHistogramOverSubscribedField
+                    = *dc.get<S_OverSubscribedField>("LocalElectronHistogramOverSubscribedField");
+                DataSpace<picongpu::simDim> const fieldGridLayoutOverSubscription
+                    = localElectronHistogramOverSubscribedField.getGridLayout().sizeWithoutGuardND();
+
+                /// @todo find better way than hard code old value, Brian Marre, 2023
+                // `static` avoids that reduce is allocating each time step memory, which will reduce the performance.
+                static pmacc::device::Reduce deviceLocalReduce = pmacc::device::Reduce(static_cast<uint32_t>(1200u));
+
+                setTimeRemaining(); // = (Delta t)_PIC
+                // fix atomic state and charge state inconsistency
+                using ForEachIonSpeciesFixAtomicState = pmacc::meta::
+                    ForEach<AtomicPhysicsIonSpecies, particles::atomicPhysics::stage::FixAtomicState<boost::mpl::_1>>;
+                ForEachIonSpeciesFixAtomicState{}(mappingDesc);
+
+                uint16_t counterSubStep = 0u;
+                // atomicPhysics sub-stepping loop, ends when timeRemaining<=0._X
+                while(true)
+                {
+                    // debug only
```

Comment:
REQUIRED: This does not look like "debug only". Remove comment or statement.

Reply:
done